From 1e01697a44f69f16da7128a0bbd65f512775bc92 Mon Sep 17 00:00:00 2001
From: Brendan Bank <brendan.bank@gmail.com>
Date: Fri, 13 Feb 2026 08:21:06 +0100
Subject: [PATCH] kea: add plugins_run hooks for DDNS plugin support

Add two hook points to allow a plugin to inject DDNS configuration:
- kea_dhcpv4_config: overlay per-subnet DDNS params into kea-dhcp4.conf
- kea_ddns_generate: generate kea-dhcp-ddns.conf and toggle keactrl
Also add kea-dhcp-ddns to syslog facilities.
---
 src/etc/inc/plugins.inc.d/kea.inc                | 16 +++++++++++++++-
 .../mvc/app/models/OPNsense/Kea/KeaDhcpv4.php    | 14 ++++++++++++++
 2 files changed, 29 insertions(+), 1 deletion(-)

diff --git a/src/etc/inc/plugins.inc.d/kea.inc b/src/etc/inc/plugins.inc.d/kea.inc
index 654387e12..957639af8 100644
--- a/src/etc/inc/plugins.inc.d/kea.inc
+++ b/src/etc/inc/plugins.inc.d/kea.inc
@@ -160,6 +160,20 @@ function kea_configure_do($verbose = false)
             $keaDhcpv6->generateConfig();
         }
         (new \OPNsense\Kea\KeaCtrlAgent())->generateConfig();
+        /* let plugins generate supplementary configs (e.g. kea-dhcp-ddns.conf) */
+        $ddnsResults = plugins_run('kea_ddns_generate');
+        $ddnsEnabled = !empty($ddnsResults);
+        /* enable/disable DDNS daemon in keactrl.conf */
+        $keactrl = '/usr/local/etc/kea/keactrl.conf';
+        $keactrlContent = @file_get_contents($keactrl);
+        if ($keactrlContent !== false) {
+            $keactrlContent = preg_replace(
+                '/^dhcp_ddns=.*$/m',
+                'dhcp_ddns=' . ($ddnsEnabled ? 'yes' : 'no'),
+                $keactrlContent
+            );
+            @file_put_contents($keactrl, $keactrlContent);
+        }
         if ($keaDhcpv6->isEnabled()) {
             mwexecfb(
                 '/usr/local/opnsense/scripts/kea/kea_prefix_watcher.py %s',
@@ -175,7 +189,7 @@ function kea_configure_do($verbose = false)
 function kea_syslog()
 {
     $logfacilities = [];
-    $logfacilities['kea'] = ['facility' => ['kea-dhcp4', 'kea-dhcp6', 'kea-ctrl-agent']];
+    $logfacilities['kea'] = ['facility' => ['kea-dhcp4', 'kea-dhcp6', 'kea-ctrl-agent', 'kea-dhcp-ddns']];
     return $logfacilities;
 }
 
diff --git a/src/opnsense/mvc/app/models/OPNsense/Kea/KeaDhcpv4.php b/src/opnsense/mvc/app/models/OPNsense/Kea/KeaDhcpv4.php
index f7f8e2164..615a6e5c0 100644
--- a/src/opnsense/mvc/app/models/OPNsense/Kea/KeaDhcpv4.php
+++ b/src/opnsense/mvc/app/models/OPNsense/Kea/KeaDhcpv4.php
@@ -292,6 +292,20 @@ class KeaDhcpv4 extends BaseModel
                 $cnf['Dhcp4']['hooks-libraries'][] = $record;
             }
         }
+        /* allow plugins to overlay config (e.g. DDNS parameters) */
+        foreach (plugins_run('kea_dhcpv4_config') as $overlay) {
+            if (isset($overlay['global'])) {
+                $cnf['Dhcp4'] = array_merge($cnf['Dhcp4'], $overlay['global']);
+            }
+            if (isset($overlay['subnets'])) {
+                foreach ($cnf['Dhcp4']['subnet4'] as &$subnet) {
+                    if (isset($overlay['subnets'][$subnet['subnet']])) {
+                        $subnet = array_merge($subnet, $overlay['subnets'][$subnet['subnet']]);
+                    }
+                }
+                unset($subnet);
+            }
+        }
         File::file_put_contents($target, json_encode($cnf, JSON_PRETTY_PRINT), 0600);
     }
 }
-- 
2.50.1 (Apple Git-155)

